### PR TITLE
Move settings endpoints to a JSON RPC server.

### PR DESCRIPTION
**Pull request by dzeckelev**
Resolves BV-769.

### CODE REVIEW

File: ui/settings.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+		err = tx.FindByPrimaryKeyTo(&settingFromDB,
+			setting.Key)
+		if err != nil {
+			if err != reform.ErrNoRows {
+				logger.Error(err.Error())
+				return nil, ErrInternal
+			}
+		}
+
+		// if setting exists and settings.permissions != 2
```
Comment (furkhat) [inline ui/settings.go]: Why break abstraction. Replace `2` with `data.ReadWrite`. Little thing but anyway.

File: ui/db.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+	"github.com/privatix/dappctrl/util/log"
+)
+
+// Transactional funcs.
+
+func (h *Handler) begin(logger log.Logger) (*reform.TX, error) {
+	tx, err := h.db.Begin()
+	if err != nil {
+		logger.Error(fmt.Sprintf(
```
Comment (ababo) [inline ui/db.go]: Suggest to replace with `"failed to begin transaction: "+err.Error()` or just `err.Error()`.

File: ui/misc.go [hunk 3]
```diff
@@ -28,6 +28,29 @@ func (h *Handler) findByPrimaryKey(logger log.Logger,
...
+			return notFoundError
+		}
+		return ErrInternal
+	}
+	return nil
+}
+
+func (h *Handler) selectAllFrom(logger log.Logger, view reform.View,
+	tail string, args ...interface{}) ([]reform.Struct, error) {
+	rows, err := h.db.SelectAllFrom(view, tail, args...)
```
Comment (ababo) [inline ui/misc.go]: Bad name (`rows`) since it can be confused with `sql.Rows`, please use `recs` instead.

File: ui/settings.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	"github.com/privatix/dappctrl/data"
+)
+
+const (
+	settingsCondition = "WHERE permissions > 0"
+	settingsUpdated   = "updated."
+)
+
+// GetSettingsResult is a GetSettings result.
+type GetSettingsResult []reform.Struct
```
Comment (ababo) [inline ui/settings.go]: Please use structs for flexibility. Already told you that.

File: ui/settings.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+	logger := h.logger.Add("method", "GetSettings", "key", key)
+
+	err := h.checkPassword(logger, password)
+	if err != nil {
+		return nil, err
+	}
+
+	if key != "" {
+		var setting data.Setting
+		err = h.selectOneTo(logger, ErrSettingNotFound, &setting,
```
Comment (ababo) [inline ui/settings.go]: Does it make sense to use `data.ReadSetting`?

File: ui/settings.go (hunk 4 above)
Comment (ababo) [reply ui/settings.go]: Please consider using `map[string]string` instead.

File: ui/settings.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+// GetSettings returns settings.
+func (h *Handler) GetSettings(password, key string) (GetSettingsResult, error) {
+	logger := h.logger.Add("method", "GetSettings", "key", key)
+
+	err := h.checkPassword(logger, password)
+	if err != nil {
+		return nil, err
+	}
+
+	if key != "" {
```
Comment (ababo) [inline ui/settings.go]: Please consider returning all the settings at once without filtering. I doubt the request-per-each-setting scenario is reasonable.

File: ui/settings.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return settings, err
+}
+
+// UpdateSettings updates settings.
+func (h *Handler) UpdateSettings(password string,
+	items []data.Setting) (*UpdateSettingsResult, error) {
```
Comment (ababo) [inline ui/settings.go]: I guess we should pass `map[string]string` instead. Indeed there's no need to update setting descriptions, etc. I also think we need to change the `data.Setting` structure (to drop non-localisable description, etc.) but this is out of scope of this task.

File: ui/db.go [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+	"fmt"
+
+	"gopkg.in/reform.v1"
+
+	"github.com/privatix/dappctrl/util/log"
+)
+
+// Transactional funcs.
+
+func (h *Handler) begin(logger log.Logger) (*reform.TX, error) {
```
Comment (ababo) [inline ui/db.go]: Consider using `InTransaction` instead of `Begin`/`Rollback`/`Commit`.

File: ui/settings.go (hunk 6 above)
Comment (dzeckelev) [reply ui/settings.go]: This request can not be performed without filtering.
Here we get all the results:
https://github.com/Privatix/dappctrl/blob/2fb2a676dcd90514a6f41056593d2ee7883b0617/ui/settings.go#L41

File: ui/settings.go (hunk 7 above)
Comment (dzeckelev) [reply ui/settings.go]: @ababo updated.

File: doc/ui/rpc.md [hunk 9]
```diff
@@ -19,6 +19,33 @@ This document describes a UI JSON RPC API located in "ui" namespace.
...
+2. Setting key (string)
+
+*Result (object)*:
+- `settings` (array of `data.Setting` objects) - settings objects.
+
+#### Update Settings
+
+*Method*:	`updateSettings`
+
+*Description*: Update exist settings.
```
Comment (ababo) [inline doc/ui/rpc.md]: existing

File: doc/ui/rpc.md [hunk 10]
```diff
@@ -19,6 +19,33 @@ This document describes a UI JSON RPC API located in "ui" namespace.
...
+
+*Parameters*:
+1. Password (string)
+2. Items maps setting keys to setting values (object)
+
+*Result (object)*:
+- `message` (string) - message about a result of the operation.
```
Comment (ababo) [inline doc/ui/rpc.md]: Did you mean error? It's always returned.

File: ui/settings.go [hunk 11]
```diff
@@ -0,0 +1,110 @@
+package ui
+
+import (
+	"gopkg.in/reform.v1"
+
+	"github.com/privatix/dappctrl/data"
+)
+
+const (
```
Comment (ababo) [inline ui/settings.go]: I would suggest to remove the whole block.

File: ui/settings.go [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+	SettingsUpdated = "updated."
+)
+
+// GetSettingsResult is a GetSettings result.
+type GetSettingsResult struct {
+	Settings []data.Setting `json:"settings"`
+}
+
+// UpdateSettingsResult is an UpdateSettings result.
+type UpdateSettingsResult struct {
```
Comment (ababo) [inline ui/settings.go]: Please drop this type if I understood correctly.

File: ui/settings.go [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+			if err != nil {
+				logger.Error(err.Error())
+				return ErrInternal
+			}
+		}
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
```
Comment (ababo) [inline ui/settings.go]: You cannot return this error directly (only our errors can be returned). I suggest to use the transaction helpers written by @furkhat.

File: doc/ui/rpc.md [hunk 14]
```diff
@@ -19,6 +19,32 @@ This document describes a UI JSON RPC API located in "ui" namespace.
...
+
+#### Get Settings
+
+*Method*:	`getSettings`
+
+*Description*: Get settings.
+
+*Parameters*:
+1. Password (string)
+2. Setting key (string)
```
Comment (ababo) [inline doc/ui/rpc.md]: We should be able to retrieve all the settings at once as `map[string]string`. Reading a single setting doesn't make much sense. Moreover single object can be retrieved by out generic method.

File: doc/ui/rpc.md [hunk 15]
```diff
@@ -19,6 +19,32 @@ This document describes a UI JSON RPC API located in "ui" namespace.
...
+
+#### Update Settings
+
+*Method*:	`updateSettings`
+
+*Description*: Update existing settings.
+
+*Parameters*:
+1. Password (string)
+2. Items maps setting keys to setting values (object)
```
Comment (ababo) [inline doc/ui/rpc.md]: Object with keys as setting names and values as setting values (object).

File: doc/ui/rpc.md [hunk 16]
```diff
@@ -96,6 +122,51 @@ This document describes a UI JSON RPC API located in "ui" namespace.
...
+
+#### Create Template
+
+*Method*:	`createTemplate`
+
+*Description*: Create new template.
+
+*Parameters*:
+1. Password (string)
+2. Template (data.Template object)
```
Comment (ababo) [inline doc/ui/rpc.md]:
```
2. Template (`data.Template` object)
```

File: doc/ui/rpc.md (hunk 14 above)
Comment (ababo) [reply doc/ui/rpc.md]: @sofabeat I think we need to reform our settings a bit. I don't see much sense having both name and key fields, why not to have just a name as PK? Moreover we don't need a description since we support localisation, so this part is going to UI's responsibility (if needed). For now I would propose just to throw non-needed fields out of `data.Setting`. Regarding RPC API I would support only two calls: `getSettings` and `updateSettings`. The first one just returns a map name->value (JSON object), the second takes such an object and apply the values (honouring permissions). What do you think?
